### PR TITLE
Make training CUDA kernels to adhere established code structure patterns

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**:
Prevent .CU kernel files include CPU headers and separate .CU kernel interfaces into separate header files as required.
Remove headers that are not even needed.

**Motivation and Context**
Our main CPU code compiles under C++14, but we still need to support CUDA 10.2 nvidia compiler that does not support C++14.

Existing training kernels include C++14 CPU only files and that fails compilation when we attempt to code in C++14, refactor or do any other changes in common infrastructure code or any other code we do not even think needs to be included into CUDA kernels. The reason for that is those kernels implementation does not conform to the established patterns that provide clear boundaries for code separation.

### CODE REVIEW

**Review by weixingzhang — DISMISSED** [review]

**Review by iK1D — APPROVED** [review]

File: orttraining/orttraining/training_ops/cuda/math/scale_impl.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  void operator()(const Tensor* scale, float& scale_value) const {
+    ORT_ENFORCE(scale->Shape().Size() == 1, "Scale input should have a single value.");
+    scale_value = static_cast<float>(*(scale->template Data<ScaleT>()));
+    ORT_ENFORCE(scale_value != 0.0f, "Scale value must not be 0.");
+  }
+};
+
+template <typename T>
+void Impl_Scale(
+    cudaStream_t stream,
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/math/scale_impl.h:20]: >cudaStream_t [](http://example.com/codeflow?start=4&length=12)

general comment: is it possible to include the header that defines cudaStream_t in these _impl headers too?

File: orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+namespace onnxruntime {
+namespace cuda {
+// Implementation can be found in cuda file, optimizers_impl.cu
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.h:8]: >optimizers_impl.cu [](http://example.com/codeflow?start=45&length=18)

is it in gradient_control_impl.cu?

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb_impl.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+namespace onnxruntime {
+namespace cuda {
+// Implementation can be found in cuda file, optimizers_impl.cu
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb_impl.h:8]: >optimizers_impl.cu [](http://example.com/codeflow?start=45&length=18)

is it in lamb_impl.cu?

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb_impl.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    float epsilon,
+    float max_norm,
+    float alpha_correction,
+    float beta_correction,
+    T2* update_direction,
+    T3* moment_1_out,
+    T3* moment_2_out,
+    size_t count);
+
+// Implementation can be found in cuda file, optimizers_impl.cu
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb_impl.h:33]: >optimizers_impl.cu [](http://example.com/codeflow?start=45&length=18)

lamb_impl.cu?

File: orttraining/orttraining/training_ops/cuda/optimizer/sg_impl.cu [hunk 5]
```diff
@@ -4,7 +4,7 @@
 #include "core/providers/cuda/cuda_common.h"
 #include "core/providers/cuda/cu_inc/common.cuh"
 #include "core/providers/cuda/atomic/common.cuh"
-#include "sg.h"
+#include "sg_impl.h"
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/optimizer/sg_impl.cu:7]: general comment: put the related header first
[https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)

**Comment by edgchen1** [comment]
> #include "core/providers/cuda/multi_tensor/common.cuh"

should we move this out of reduction_all.h?

---
Refers to: orttraining/orttraining/training_ops/cuda/reduction/reduction_all.h:6 in 222cc80. [](commit_id = 222cc80b5f7f637eadb41c789951742207b1121d, deletion_comment = False)

File: orttraining/orttraining/training_ops/cuda/math/scale_impl.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+#pragma once
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename ScaleT>
+struct GetScaleValueImpl {
+  void operator()(const Tensor* scale, float& scale_value) const {
+    ORT_ENFORCE(scale->Shape().Size() == 1, "Scale input should have a single value.");
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply orttraining/orttraining/training_ops/cuda/math/scale_impl.h:12]: I think I moved this one by mistake

File: orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.h (hunk 2 above)
Comment (yuslepukhin) [reply orttraining/orttraining/training_ops/cuda/optimizer/gradient_control_impl.h:8]: I just moved this comment. I will fix it.

File: orttraining/orttraining/training_ops/cuda/math/scale_impl.h (hunk 1 above)
Comment (yuslepukhin) [reply orttraining/orttraining/training_ops/cuda/math/scale_impl.h:20]: I hesitate doing it, getting into the headers nightmare.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb_impl.h (hunk 3 above)
Comment (yuslepukhin) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb_impl.h:8]: Yes

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb_impl.h (hunk 4 above)
Comment (yuslepukhin) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb_impl.h:33]: yes

File: orttraining/orttraining/training_ops/cuda/math/scale_impl.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+namespace onnxruntime {
+namespace cuda {
+
+#include <cuda_runtime.h>
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/math/scale_impl.h:9]: nit: move out of onnxruntime namespace

File: onnxruntime/contrib_ops/cuda/math/isfinite_impl.h [hunk 8]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (edgchen1) [inline onnxruntime/contrib_ops/cuda/math/isfinite_impl.h:1]: noticed that contrib isfinite deviates from the pattern of x.h, x.cc, x_impl.h, x_impl.cu
why is that?

File: orttraining/orttraining/training_ops/cuda/math/isfinite.h [hunk 9]
```diff
@@ -4,7 +4,7 @@
 #pragma once
 #include "core/common/common.h"
 #include "core/providers/cuda/cuda_kernel.h"
-#include "core/providers/cuda/multi_tensor/common.cuh"
+// #include "core/providers/cuda/multi_tensor/common.cuh"
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/math/isfinite.h:7]: remove commented line?

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**